### ISSUE TITLE
findOne() return undefined

### ISSUE THREAD

**Issue by fwh1990** [typeorm/typeorm#1865]
**Issue type:**

[ ] bug report

**Database system/driver:**

[ ] `mysql`

**TypeORM version:**

[ ] `@next`

**Steps to reproduce or a small repository showing the problem:**

```js
// id === 1
// xx === null
// yy === null

 // return undefined
await getConnection().getRepository(User).findOne({
    select: ['xx'],
    where: `m = 1`,
});

// return undefined
await getConnection().getRepository(User).findOne({
    select: ['xx', 'yy'],
    where: `id = 1`,
});

// return User instance
await getConnection().getRepository(User).findOne({
    select: ['id', 'xx', 'yy'],
    where: `id = 1`,
}); 
```

It means one of item in select should not be null.

**Comment by pleerock** [comment]
id is required for selection

**Comment by fwh1990** [comment]
No. If column `xx` has value or `xx === ''`, findOne will return object result like:
```js
{
  xx: ''
}

{
  xx: 'value in db'
}
```

**Comment by fwh1990** [comment]
You should read my last words:

`It means one of item in select should not be null.`

**Comment by pleerock** [comment]
I still did not get you. You wrote two objects in `return` which one is returned you?

**Comment by fwh1990** [comment]
```js
// xx === null
await getConnection().getRepository(User).findOne({
    select: ['xx'],
    where: `id = 1`,
});

// return 
null
```

```js
// xx === ''
await getConnection().getRepository(User).findOne({
    select: ['xx'],
    where: `id = 1`,
});

// return 
{
  xx: ''
}
```

```js
// xx === 'cool'
await getConnection().getRepository(User).findOne({
    select: ['xx'],
    where: `id = 1`,
});

// return 
{
  xx: 'cool'
}

```

**Comment by homj** [comment]
So I just experienced this issue as well. And I cannot wrap my head around why findOne behaves like this.
When you select a single attribute in a `findOne` call and this attribute is `null`, the whole result is `null`.

So as @fwh1990 explained, for a call like this:

```typescript
const result = await repository.findOne({
      where: {
          year: year
      },
      select: [ 'someNullableColumn' ]
  });
````

result will be `undefined` instead of `{ someNullableColumn: null }`.
For us, there's a significant difference between the whole result being `undefined` (which **should** mean there is no record matching the `where` clause) and that single column being `null`

**Comment by eliastg** [comment]
Same problem here.
copy/paste from documentation, and still get undefined when I use findOne, to find a user by its email.
In my case I am using NESTJS, but TypeOrm is referenced in NESTJS documentation for this.
What am I doing wrong?

`
export abstract class BaseEntity {
    @PrimaryGeneratedColumn('uuid')
    id: string;

    @CreateDateColumn({ type: 'timestamptz', default: () => 'CURRENT_TIMESTAMP' })
    createdAt: Date;

    @UpdateDateColumn({ type: 'timestamptz', default: () => 'CURRENT_TIMESTAMP' })
    updatedAt: Date;

    @DeleteDateColumn({ type: 'timestamptz', default: () => 'CURRENT_TIMESTAMP' })
    deletedAt: Date;
}

@Entity()
export class User extends BaseEntity {
    @Column({ nullable: false })
    fullName: string;

    @Column({ nullable: false, unique: true })
    email: string;

    @Column({ nullable: false })
    password: string;
}

@EntityRepository(User)
export class UserRepository extends Repository<User> {
    
}

@Injectable()
export class UserService {
    private readonly logger = new Logger(UserService.name, { timestamp: true });

    constructor(
        private readonly repository: UserRepository
    ){}

    createOne( userData: CreateUserDto): Observable<User> {
        const newUser: User = {
            ...this.repository.create(),
            ...userData
        };

        // this.logger.debug(`Cretaing User with this data: ${JSON.stringify(userData)}`);
        return from(this.repository.save(newUser));
    }

    findOneByEmail(email: string): Observable<User> {
        return from(this.repository.findOne({ email }));
    }
}

**Comment by joss12** [comment]
I had the same issue and i just fixed it. there is my solution
  const user = await repository.findOne({
    select:['firstName', 'lastName', 'email'],
    where: payload.id
  }); If you only want to show the informations entered(firstName, lastName, email)

or you can still use select:[empty] you will get the entity
